File: day08/src/main/java/de/havox_design/aoc2017/day08/Day08.java
package de.havox_design.aoc2017.day08;

import de.havox_design.aoc2017.utils.DataReader;

import java.util.List;

public class Day08 {
    private final List<String> input;

    public Day08(String fileName) {
        input = readData(fileName);
    }

    public static long solvePart1(String fileName) {
        Day08 instance = new Day08(fileName);
        return instance.solvePart1();
    }

    public static long solvePart2(String fileName) {
        Day08 instance = new Day08(fileName);
        return instance.solvePart2();
    }

    public long solvePart1() {
        return 0L;
    }

    public long solvePart2() {return 0L;
    }

    private List<String> readData(String fileName) {
        return DataReader.readData(fileName, MainClass.class);
    }
}

File: day08/src/main/scala/de/havox_design/aoc2017/day08/Day08.scala
package de.havox_design.aoc2017.day08

object Day08 {
  def solvePart1(filename: String): Int =
    compute(readData(filename)
      .toIndexedSeq)
      ._1
      .values
      .max

  def solvePart2(filename: String): Int =
    compute(readData(filename)
      .toIndexedSeq)
      ._2

  def main(args: Array[String]): Unit = {
    def dayFileName = "day08.txt"

    println("Solution for part1: " + solvePart1(dayFileName))
    println("Solution for part2: " + solvePart2(dayFileName))
  }

  def compute(input: Seq[String]): (Map[String, Int], Int) =
    val initial = Map
      .empty[String, Int]
      .withDefaultValue(0) -> 0

    input
      .map(_.split(" "))
      .foldLeft(initial) { case ((registers, max), instruction) =>
        val Array(destination, operation, amount, _, test, condition, threshold) = instruction
        val left = registers(test)
        val right = threshold
          .toInt
        val valid = condition match
          case ">" => left > right
          case "<" => left < right
          case ">=" => left >= right
          case "<=" => left <= right
          case "==" => left == right
          case "!=" => left != right

        val modifier = if !valid then {
          0
        }
        else {
          if operation != "inc" then {
            -1
          }
          else {
            1
          }
        }

        val next = registers(destination) + amount.toInt * modifier
        registers
          .updated(destination, next) -> max
          .max(next)
      }

  private def readData(filename: String): Iterator[String] =
    scala
      .io
      .Source
      .fromResource(filename)
      .getLines()
}
